### ISSUE TITLE
Tensorflow FIFOQueueV2 unknown layer type error

### ISSUE THREAD

**Issue by lasandcris** [opencv/opencv#9137]
##### System information (version)
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015

##### description

I'm using the tf_importer.cpp to import a tensorflow pb frozen model. However, I get an error during populating to the Net class variable: 

batch/fifo_queue:FIFOQueueV2
container:[ ]
component_types:[ ]
shapes:[ ]
capacity:[ ]
shared_name:[ ]
OpenCV Error: Unspecified error (Unknown layer type FIFOQueueV2 in op batch/fifo_queue) in cv::dnn::experimental_dnn_v1::`anonymous-namespace'::TFImporter::populateNet, file ..\..\..\source\modules\dnn\src\tensorflow\tf_importer.cpp, line 735

Is there a way I can add this layer type into tf_importer.cpp? Or how could I remove this as this is not used in the network any longer after training. 

thanks

**Comment by dkurt** [comment]
@lasandcris, could you try to apply a one more procedure over the frozen model? See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference.py.

**Comment by lasandcris** [comment]
@dkurt, thanks for your suggestion, I'll give it a try in a minute. However, I can already see the next Layer being a problem and that is RestoreV2. Have you got any tips for that?

**Comment by dkurt** [comment]
It'll be much easier to say if you share this model with us. Is it possible? Please upload your `.pb` somewhere.

**Comment by lasandcris** [comment]
@dkurt, sure, this is the link to the model:
[](https://drive.google.com/file/d/0B3yxfi3uSufjQkxWNXNyRkdySU0/view?usp=sharing)

**Comment by lasandcris** [comment]
@dkurt, have you managed to download the model?

**Comment by dkurt** [comment]
Yes, thank you! I've just started work with it.

**Comment by dkurt** [comment]
@lasandcris , model seems to me a kind of strange. I've tried to load it by
```python
with tf.gfile.FastGFile('frozen_model.pb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    for node in graph_def.node:
        print node.op
```
and output is only
```bash
Const
Const
Const
Const
RestoreV2
Assign
```
In example, the same code for Inception-5h model @ https://github.com/opencv/opencv/blob/master/samples/dnn/tf_inception.cpp produces Const, Conv2D, Relu, MaxPool etc. Could you please check the model?

**Comment by lasandcris** [comment]
@dkurt, i was just trying to save on downloading a large file. this is the full model:

https://drive.google.com/file/d/0B3yxfi3uSufjVzRjOUhaNVdxT2c/view?usp=sharing

and this is the graph:

https://drive.google.com/file/d/0B3yxfi3uSufjSzB0cXl3YXNUUHc/view?usp=sharing

thanks

**Comment by lasandcris** [comment]
hi @dkurt, have you managed to check out the model? 
thanks

**Comment by dkurt** [comment]
@lasandcris, I've tried to apply [optimize_for_inference.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference.py) as mentioned above:
```bash
./bazel-bin/tensorflow/python/tools/optimize_for_inference \
  --input ~/Downloads/frozen_model_full.pb \
  --output ~/Downloads/frozen_model_full_opt.pb \
  --frozen_graph=True \
  --input_names="batch" \
  --output_names="fake_discriminator/discriminator/layer_5/Sigmoid,real_discriminator/discriminator/layer_5/Sigmoid"
```
Could you try to freeze model in single image mode (batch size == 1)? It seems to me that it will help easily drop the first few operations are used for batching.

**Comment by lasandcris** [comment]
@dkurt , I'll have a swing at it now,
thanks

**Comment by lasandcris** [comment]
@dkurt, it now hit another layer error:

OpenCV Error: Unspecified error (Unknown layer type Pad in op generator/encoder_1/conv/Pad)

**Comment by lasandcris** [comment]
hi @dkurt ,
have you had any time to look into this?
thanks

**Comment by dkurt** [comment]
@lasandcris, I've asked you to freeze the model with batch size 1 and share it with me because I can't load the old one even in TensorFlow.

**Comment by lasandcris** [comment]
hi @dkurt, the frozen_model_full.pb is frozen with a batch size of 1 already and I imported it in tensorflow without any problem. But here you can find the original model so you can dig into it. https://drive.google.com/file/d/0B3yxfi3uSufjaHhkNFNnYWR4bG8/view?usp=sharing

**Comment by dkurt** [comment]
@lasandcris , I tried to do something with this model. Truly, there are several operations that should be considered at `tf_importer` like element wise summation (Add), absolute value activation layer (Abs), scale by constant value (Mul), one more variation of padding operation (Pad), batch normalization (is it?). And it requires just few lines of code. But the thing is this model is so complicated for testing. I faced problems like
* There are two inputs and two outputs. It's hard to say they are independent or not now.
* Even after applying `freeze_graph` + `optimize_for_inference` tools there are operations like `Mean` for batch norm, `StopGradient` and a lot of others like `SquaredDifference`, `Rsqrt`, `Square`, `Reciprocal`, etc. Please, could you get a reference to an article or some source to this model? Based just on operation types it doesn't looks like common deep learning pipeline.

I would like to solve the problem with missing operations on some public models now. After that I'll continue with this one.

**Comment by dkurt** [comment]
TensorFlow parser will be extended at https://github.com/opencv/opencv/pull/9287.
It seems to me, your model use batch normalization. It represents as sequence of atomic operations between constants (summation, multiplication, squared root), but it's just a scale and shift finally (`bn = in * w + b`).
Would it be OK with you if you save graph using `tf.contrib.layers.batch_norm(..., fused=True, is_training=False)`?
